### PR TITLE
Better managing of Mongo Updates/Tracker

### PR DESCRIPTION
**Pull request by ToyboxZach**
Restructure how we handle the Tracker, instead of a global listen to any change make the listens happen on specific queries and collections. Also fixes some bugs where we would generate way too many instances of subscriptions.

This also fixes the log in flow so the series of loggingIn -> LoggedIN is consistent and doesn't allow the package to ever auto log you out as that should be handled by the owning app just like in base meteor.

### CODE REVIEW

**Comment by ToyboxZach** [comment]
Every time you register a callback it gets added to a generic "changed" callback, that means that every single callback needs to be called on every single change to a collection which causes a huge hold up on the UI thread.

To make it even worst the subscriptions don't get cleaned up when you resubscribe so every time a "withTracker" or "useTracker" was called you would get stacked subscriptions. This then would lead to 100 plus subscriptions for a single component after some time which leads to bigger and bigger slow downs.

My PR separates that out as much as possible and makes sure that you only have as many subscriptions as you actually need. It is a pretty hefty change and I don't personally use and local collections in my app so those are going to be pretty poorly tested, but if other people want to take my change as a jumping off point I think its a pretty good point for server only subscriptions.

This also fixes up weirdness with the logging in and order of events where you would have an invalid user state or it would force you to logout because of bad internet

Likely fixes these erorrs:
https://github.com/TheRealNate/meteor-react-native/issues/75
https://github.com/TheRealNate/meteor-react-native/issues/58

and maybe
https://github.com/TheRealNate/meteor-react-native/issues/79

**Comment by TheRealNate** [comment]
Hi @ToyboxZach,

I want to thank you for your extreme amount of work on this, however I don't think we can safely merge in a PR of this size.

We're going to adjust our contribution guidelines and formatting spec shortly, then we can circle back to these changes.

Closing for now.

**Comment by adamgins** [comment]
@TheRealNate any more thoughts on an actual solution for the issue? This is a major issue for us too.

**Comment by TheRealNate** [comment]
@adamgins what are the specific issues you're seeing?

**Comment by arianjahiri** [comment]
@TheRealNate Hi, commenting on behalf of @adamgins. Essentially we were seeing the issues that were described by @ToyboxZach above and in the issues https://github.com/meteorrn/meteor-react-native/issues/75 https://github.com/meteorrn/meteor-react-native/issues/96, excessive callbacks. We were able to find this from profiling our application using the hermes profiler. We previously used our own fork of inProgress-team/react-native-meteor but now migrated to a fork of ToyboxZach/meteor-react-native as we have some custom code of our own that runs in the meteor-react-native library. Anyway, switching the bases of our  meteor-react-native fork to utilise the new meteorrn/meteor-react-native changes as well as @ToyboxZach's changes, we found that it solved our excessive callback issue. 

It is hard to reproduce this issue as we did not notice it before but have run into it now as our application has become more complex. Not sure if it will be helpful but below is the before and after of doing the same action in our application (editing a document and observing the live changes):

**Before**: Excessive callbacks occurring due to Data.js highlighted in the red boxes.
PROFILE_CALLBACK_HELL

**After**: No sign of callbacks.
PROFILE_NO_CALLBACK_HELL

**Comment by ToyboxZach** [comment]
I have no time to push this any past what I need but fully support if someone wants to use this as a jumping off point for a new npm package, but I do understand I didn’t make this change planning to merge it into this existing package and it is extremely risky for a lot of existing apps that use features that I do not.

**Comment by bratelefant** [comment]
Just wanted to quickly report that @ToyboxZach s fork solves all my performance issues with useTracker from the core package immediately. Had two simple trackers each of them subscribing to a meteor publication. The subscriptions got called up to 15 times, resulting in massive lags of the ui thread and bursting the server, although dependency list was empty.

I'd really love to get this merged.

**Comment by adamgins** [comment]
@TheRealNate is it possible to please  re-assess the position there's too much code to merge, given the impact here?

**Comment by ToyboxZach** [comment]
I believe if someone has time to push it through you could split my change into a few things:

src/components/useTracker.js This should solve most of the problem of too many trackers being added and not being cleaned up, I believe you can take that file as a single change and it will improve a chunk of these bugs

src/user/User.js I was having trouble with it autologging out or spamming my server too much, I did my best to fix the reactivity ( I know this isn't perfect as is, but it was a huge improvement experience-wise for me)

src/Collection.js src/Data.js src/Meteor.js src/ReactiveDict.js
These changes are to fix up the issue of too many things being updated at once, so it splits the the updates out to be much more specific, similar to how web handles it

Everything else appears to be prettifying or some change I must have pulled in elsewhere.

Again sorry I don't have the time to break these into smaller testable PRs myself

**Comment by bratelefant** [comment]
One up for the useTracker file.

**Comment by TheRealNate** [comment]
@ToyboxZach yeah agreed the majority of the changes are prettier changes that make it harder to review.

@jankapunkt, what are your thoughts on merging this?

Reopening

**Comment by bratelefant** [comment]
Tried to get a minimal diff using @ToyboxZach s PR to resolve the useTracker issues. Basically step by step merged changes that seemed significant to the solution, did not yet review the code changes in detail, but the seem to be solved at the first glance. Will do some further testing. Please cf my fork.

**Comment by jankapunkt** [comment]
hey @TheRealNate @ToyboxZach I can take a look at it. @ToyboxZach can you please comment on the files / lines that are most related to the fixes? I would go for them first and go through the code review etc. afterwards.

Please also note, that I have updated minimongo and removed nearly all deprecated dependencies that were replaceable by native es6+ code. So we somehow need to merge both of them into master.

File: lib/ddp.js [hunk 1]
```diff
@@ -23,6 +21,47 @@ const PUBLIC_EVENTS = [
 ];
 const DEFAULT_RECONNECT_INTERVAL = 10000;
 
+class EventInterface {
```
Comment (ToyboxZach) [inline lib/ddp.js:24]: Don't remember if I did this or not, but this file is definitely important for reactivity

File: src/Collection.js [hunk 2]
```diff
@@ -8,30 +8,52 @@ import call from './Call';
...
-  if(observers[collection]) {
-    observers[collection].forEach(({cursor, callbacks}) => {
-      if(callbacks[type]) {
-        if(type === 'removed') {
-          callbacks['removed'](newDocument);
-        }
-        else if(Data.db[collection].findOne({$and:[{_id:newDocument._id}, cursor._selector]})) {
+const observersByComp = {};
+
+export function getObservers(type, collection, newDocument) {
```
Comment (ToyboxZach) [inline src/Collection.js:13]: This function is important for reactivity fix

File: src/Meteor.js [hunk 3]
```diff
@@ -36,7 +37,7 @@ const Meteor = {
   },
   status() {
     return {
-      connected: Data.ddp ? Data.ddp.status == 'connected' : false,
+      connected: !!this._reactiveDict.get('connected'),
```
Comment (ToyboxZach) [inline src/Meteor.js:40]: Changes with the reactiveDict are important for reactivity change

File: src/Meteor.js [hunk 4]
```diff
@@ -91,44 +102,60 @@ const Meteor = {
...
+            Data.ddp.connect();
+          }
         }
-      });
-    }
-    catch(e) {
-      console.warn("Warning: NetInfo not installed, so DDP will not automatically reconnect");
+      );
+
+      NetInfo.fetch().then(
```
Comment (ToyboxZach) [inline src/Meteor.js:116]: Important to tell if connected or not for reactivity

File: src/Meteor.js [hunk 5]
```diff
@@ -91,44 +102,60 @@ const Meteor = {
...
-      Data.notify('change');
-
-      if(isVerbose) {
+      if (isVerbose) {
         console.info('Connected to DDP server.');
       }
       this._loadInitialUser().then(() => {
         this._subscriptionsRestart();
       });
+      this._reactiveDict.set('connected', true);
```
Comment (ToyboxZach) [inline src/Meteor.js:146]: Important to be reactive to connected

File: src/Meteor.js [hunk 6]
```diff
@@ -91,44 +102,60 @@ const Meteor = {
...
+      this.connected = true;
+      Data.notify('change');
     });
 
     let lastDisconnect = null;
     Data.ddp.on('disconnected', () => {
+      this.connected = false;
```
Comment (ToyboxZach) [inline src/Meteor.js:153]: Important to be reactive to connected

File: src/Meteor.js [hunk 7]
```diff
@@ -151,8 +178,14 @@ const Meteor = {
       };
 
       Data.db[message.collection].upsert(document);
-
-      runObservers("added", message.collection, document);
+      let observers = getObservers('added', message.collection, document);
+      observers.forEach(callback => {
```
Comment (ToyboxZach) [inline src/Meteor.js:182]: important to improve reactivity of collections

File: src/Meteor.js [hunk 8]
```diff
@@ -180,26 +213,47 @@ const Meteor = {
...
-      if(Data.db[message.collection]) {
+      if (Data.db[message.collection]) {
         const document = {
           _id: message.id,
           ...message.fields,
           ...unset,
         };
 
-        const oldDocument = Data.db[message.collection].findOne({_id:message.id});
+        const oldDocument = Data.db[message.collection].findOne({
```
Comment (ToyboxZach) [inline src/Meteor.js:223]: Important to fix reactitvity of collections

File: src/Meteor.js [hunk 9]
```diff
@@ -180,26 +213,47 @@ const Meteor = {
...
+          }
+        });
       }
     });
 
     Data.ddp.on('removed', message => {
-      if(Data.db[message.collection]) {
-        const oldDocument = Data.db[message.collection].findOne({_id:message.id});
+      if (Data.db[message.collection]) {
+        const oldDocument = Data.db[message.collection].findOne({
```
Comment (ToyboxZach) [inline src/Meteor.js:241]: Important to fix reactitvity of collections

File: src/Meteor.js [hunk 10]
```diff
@@ -217,6 +271,9 @@ const Meteor = {
         }
       }
     });
+    Data.ddp.on('error', message => {
```
Comment (ToyboxZach) [inline src/Meteor.js:274]: Helpful for debugging

File: src/Meteor.js [hunk 11]
```diff
@@ -338,6 +395,10 @@ const Meteor = {
           }
         });
       });
+    } else {
+      if (Data.subscriptions[id]) {
```
Comment (ToyboxZach) [inline src/Meteor.js:399]: Important to fix reactitvity of collections

File: src/ReactiveDict.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import EJSON from 'ejson';
```
Comment (ToyboxZach) [inline src/ReactiveDict.js:1]: This file is Important to fix reactitvity of collections and connected

File: src/components/useTracker.js [hunk 13]
```diff
@@ -1,30 +1,50 @@
-import { useEffect, useState } from 'react';
+import {
```
Comment (ToyboxZach) [inline src/components/useTracker.js:1]: Important to fix duplicate trackers

File: src/user/Accounts.js [hunk 14]
```diff
@@ -6,21 +6,22 @@ import Meteor from '../Meteor.js';
 
 class AccountsPassword {
   _hashPassword = hashPassword;
-  
+
   createUser = (options, callback = () => {}) => {
     // Replace password with the hashed password.
     options.password = hashPassword(options.password);
```
Comment (ToyboxZach) [inline src/user/Accounts.js:13]: Improves log-in experience and prevents auto logouts

File: src/user/User.js [hunk 15]
```diff
@@ -2,28 +2,35 @@ import Data from '../Data';
 import { hashPassword } from '../../lib/utils';
```
Comment (ToyboxZach) [inline src/user/User.js:2]: Improves log-in experience and prevents auto logouts

**Comment by ToyboxZach** [comment]
A lot of these changes are only showing up here because its pulling into dev, which is behind master by a chunk of commits. A PR directly to master would be much smaller

**Comment by ToyboxZach** [comment]
This one is much more sane, still has a bunch of prettify-ing which again sorry, but its much fewer changes
https://github.com/meteorrn/meteor-react-native/pull/93/files?diff=split&w=1

**Comment by TheRealNate** [comment]
> A lot of these changes are only showing up here because its pulling into dev, which is behind master by a chunk of commits. A PR directly to master would be much smaller or if dev is updated to master

@ToyboxZach Ah gotcha. I've brought dev up to date with master. Could you try rebasing and lets see how big the PR ends up being?

**Comment by ToyboxZach** [comment]
I messed up a little bit while rebasing, and misunderstood a prompt, I think I have everything back to what I had before

**Comment by jankapunkt** [comment]
Rebase can be such a pain... Let me know when you won the battle. Alternatively, there will be no blame if you close and open a new PR in case this gets out of hand.

**Comment by jankapunkt** [comment]
@ToyboxZach could you please take a look at #99 and see if there may be potential clashes? I am still trying to figure out how we get both PRs work out in a smooth fashion.

**Comment by ToyboxZach** [comment]
My suggestion would be to make a change that only updates prettier and runs it on all the files, then merge that in, then I can load in that change, pretty much ignoring all of it, and then run prettier again in the same fashion so that most things don't change

**Comment by jankapunkt** [comment]
@ToyboxZach sorry for late response :see_no_evil: I agree on that as it makes much sense. I would therefore go with #99 (it also includes all the prettier updates etc.) and then we tackle this one, right?

**Comment by jankapunkt** [comment]
@ToyboxZach I merged #99 it's now on dev and contains all prettier fixes. It also contains CI for prettier lint and unit tests. You can merge dev and then continue providing the changes. A review from my end should be much easier without the prettier changes.

**Comment by ToyboxZach** [comment]
I opened a new PR from a different branch so that I don't mess up my main up. Its still probably two PRs within one, and I don't have the time to break it apart at all, but should be easier to see what I'm doing now

**Pull request closed without merging**